README.md
# ember-stagger-swagger

[![npm version](https://badge.fury.io/js/ember-stagger-swagger.svg)](https://badge.fury.io/js/ember-stagger-swagger)
[![Build Status](https://travis-ci.org/BrianSipple/ember-stagger-swagger.svg?branch=master)](https://travis-ci.org/BrianSipple/ember-stagger-swagger) [![Ember Observer Score](http://emberobserver.com/badges/ember-stagger-swagger.svg)](http://emberobserver.com/addons/ember-stagger-swagger)

*GPU stagger animation for Ember list items*

![](http://33.media.tumblr.com/29addca2c908d96a071932761ffd177a/tumblr_nstg1jgKcg1uruo10o1_500.gif)

See the demo [here][demo].


## Installation
```
ember install ember-stagger-swagger
```

## Usage

`ember-stagger-swagger` ships with a `stagger-set` component that can be used directly in a template to wrap the items that you wish to animate.

Conceptually, the component treats all direct child elements as its _set items_ or _list items_:

```htmlbars

<h2>Spell Ingredients </h2>

{{#stagger-set inDirection="right" inEffect="slideAndFade"}}

  {{#each potions as |potion|}}
    <li>{{potion.name}}</li>
  {{/each}}

{{/stagger-set}}

```

Additionally, `ember-stagger-swagger` exposes a mixin that can be imported directly from the addon and extended however you see fit:
```js.es6
import StaggerSetMixin from 'ember-stagger-swagger/mixins/stagger-set';
```

## API

##### <a name="api-inDirection"></a>`inDirection`  
  * _description_: The direction of animation when the items stagger into view.
  * ***_required_: yes***
  * _constraints_: A string keyword matching either `'left'`, `'down'`, `'right'`, or `'up'`.

##### <a name="api-outDirection"></a> `outDirection`  
  * _description_: The direction of animation when the items stagger out of view.
  * _required_: no
  * _default_: The provided `inDirection`.
  * _constraints_: A string keyword matching either `'left'`, `'down'`, `'right'`, or `'up'`.

##### <a name="api-inEffect"></a> `inEffect`  
  * _description_: A recognized animation effect applied to each list item when it's animating in.
  * ***_required_: yes***
  * _constraints_: A string keyword matching either `'slideAndFade'`, `'slide'`, `'fade'`, or `'scale'` (see the [demo][demo] for a preview of each).

##### <a name="api-outEffect"></a> `outEffect`  
  * _description_: A recognized animation effect applied to each list item when it's animating in.
  * _required_: no
  * _default_: the provided `inEffect`
  * _constraints_: A string keyword matching either `'slideAndFade'`, `'slide'`, `'fade'`, or `'scale'` (see the [demo][demo] for a preview of each).

##### <a name="api-customInEffect"></a> `customInEffect`  
  * _description_: An [animation name](https://developer.mozilla.org/en-US/docs/Web/CSS/animation-name) matching a [CSS animation keyframe](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes) that you have defined in your project. If specified alongside an `inEffect`, this name will take precedence. This can be used to create your own effects and integrate them with `stagger-swagger`'s built-in functionality.
  * _required_: no
  * _default_: `null`
  * _constraints_: None. Just make sure the name matches a CSS animation keyframe that you have defined in your project.

##### <a name="api-customOutEffect"></a> `customOutEffect`  
  * _description_: An [animation name](https://developer.mozilla.org/en-US/docs/Web/CSS/animation-name) matching a [CSS animation keyframe](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes) that you have defined in your project. If specified alongside an `outEffect`, this name will take precedence. This can be used to create your own effects and integrate them with `stagger-swagger`'s built-in functionality.
  * _required_: no
  * _default_: `null`
  * _constraints_: None. Just make sure the name matches a CSS animation keyframe that you have defined in your project.


##### <a name="api-showItems"></a> `showItems`  
  * _description_: Manual hook for toggling the set between its entrance and exit animations.
  * _required_: no
  * _default_: `true` to correspond with [`enterOnRender`](#api-enterOnRender)'s default.
  * _constraints_: `true` or `false`    


##### <a name="api-enterOnRender"></a> `enterOnRender`  
  * _description_: Whether or not the elements in the stagger set should animate into view when the component is rendered.  
  *Note:* `enterOnRender` allows for a
  more fine-grained level of control than just using `showItems`. Without `enterOnRender`, initializing the component with `showItems` set to `true` will cause the items to render in their normal visible state, from which the animation can be toggled further. Setting `enterOnRender` to true -- in conjunction with setting `showItems` to true (both of which are the default) -- creates a stagger-in animation on render and then hinges on the state of `showItems` going forward.
  * _required_: no
  * _default_: `true`
  * _constraints_: `true` or `false`   


##### <a name="api-staggerInterval"></a> `staggerInterval`
  * _description_: the number of milliseconds between the animation of successive items in the set.
  * _required_: no
  * _default_: 32ms. 2 frames per item (1 frame @ 60fps ~= 16ms) creates a noticeably staggered but still-perceptively [smooth and fluid motion](https://en.wikipedia.org/wiki/Traditional_animation#.22Shooting_on_twos.22).
  * _constraints_: a number value greater than or equal to 32.


##### <a name="api-inDelay"></a> `inDelay`  
  * _description_: Duration of delay for the items' entrance animation (when the animation is activated).
  * _required_: no
  * _default_: 0
  * _constraints_: a number value greater than or equal to 0.

##### <a name="api-outDelay"></a> `outDelay`  
  * _description_: Duration of delay for the items' exit animation (when the animation is activated).
  * _required_: no
  * _default_: 0
  * _constraints_: a number value greater than or equal to 0.


##### <a name="api-inTimingFunc"></a> `inTimingFunc`  
  * _description_: The [animation-timing-function](https://developer.mozilla.org/en-US/docs/Web/CSS/animation-timing-function) applied to each item in the set when it's animating in.
  * _required_: no
  * _default_: `cubic-bezier(0.215, 0.610, 0.355, 1.000)` (AKA ["ease-out-cubic"](http://easings.net/#easeOutCubic))
  * _constraints_: a string matching any [valid CSS `timing-function` value](https://developer.mozilla.org/en-US/docs/Web/CSS/timing-function). If this value is invalid, the browser will default to using [`ease`](http://cubic-bezier.com/#.25,.1,.25,1).


##### <a name="api-outTimingFunc"></a> `outTimingFunc`  
  * _description_: The [animation-timing-function](https://developer.mozilla.org/en-US/docs/Web/CSS/animation-timing-function) applied to each item in the set when it's animating out.
  * _required_: no
  * _default_: `cubic-bezier(0.55, 0.055, 0.675, 0.19)` (AKA ["ease-in-cubic"](http://easings.net/#easeInCubic))
  * _constraints_: a string matching any [valid CSS `timing-function` value](https://developer.mozilla.org/en-US/docs/Web/CSS/timing-function). If this value is invalid, the browser will default to using [`ease`](http://cubic-bezier.com/#.25,.1,.25,1).

##### <a name="api-inDuration"></a> `inDuration`  
  * _description_: The duration (in milliseconds) that *a single item* will take when animating in.
  * _required_: no
  * _default_: 500
  * _constraints_: a numeric value greater than 0

##### <a name="api-outDuration"></a> `outDuration`  
  * _description_: The duration (in milliseconds) that *a single item* will take when animating out.
  * _required_: no
  * _default_: 500
  * _constraints_: a numeric value greater than 0

##### <a name="api-duration"></a> `duration`  
  * _description_: A convenience property to set a single duration on both the entrance and exit animations. If set alongside any `inDuration` or `outDuration`, this property will take precedence
  * _required_: no
  * _default_: null (property is ignored if unset),
  * _constraints_: a numeric value greater than zero




### Practical Tips

#### Styling
Because the DOM elements of Ember components are, by default, `<div>`s, and because it handles setting an `animationName` property on the component's direct children, you can safely design, conceptualize, and style your child elements as you normally would for the list items of a relative container.

Furthermore, because the keyframes for the built-in effects of `slide` and `slideAndFade` define transforms to bring their element in or out of its container's visible bounds (e.g., `transform: translate3d(-100, 0, 0)` at the 100%-stop of a left slide), it may well be useful to restrict overflow on the top-level component's element so that the children disappear when they're outside of said bounds.

The [stagger-set "list items" demo][demo] is an example of how this would appear.

#### Creating Animation Effects
By default, a `stagger-list` component will attempt to map the keywords provided for `inEffect` or `outEffect` to one of its [built-in keyframes](/app/styles/_keyframes.css).

However, you're free to implement your own keyframes and have them called instead. Just define them in your stylesheets as you would normally, and then pass the keyframe name to a `stagger-list` as a string argument for either `customInEffect` or `customOutEffect`. When these attributes are defined, `stagger-list` will always set them on the `animation-name` property of its child elements' style definition at the appropriate time.

## Action Hooks

* `onAnimationStart`: called immediately after the _first_ item in the set triggers its [`animationstart`](https://developer.mozilla.org/en-US/docs/Web/Events/animationstart) event.
  * called with:
    * `animationEvent`: the [`animationevent` object](https://developer.mozilla.org/en-US/docs/Web/Events/animationstart#Properties)

* `onAnimationComplete`: called immediately after the _last_ item in the set triggers its [`animationend`](https://developer.mozilla.org/en-US/docs/Web/Events/animationend) event.
  * called with:
    * `animationEvent`: the [`animationevent` object](https://developer.mozilla.org/en-US/docs/Web/Events/animationend#Properties)

Together, these hooks can provide more control over interactions with the component during its animation. For example, if you set up a button to trigger toggles of the animation, you might want to make sure that it's disabled between the start and completion events.

```htmlbars

<h2>Spell Ingredients </h2>

{{#stagger-set
  inDirection="right"
  inEffect="slideAndFade"
  onAnimationStart=(action 'loadCannons')
  onAnimationComplete=(action 'fireCannons')
}}

  {{#each potions as |potion|}}
    <li>{{potion.name}}</li>
  {{/each}}

{{/stagger-set}}

```


## Future Goals & Ideas
* _Tentative_: [Explore improved effects for the vertical `slide` animation](#link-to-issue)?  
  * possibly break the mixins apart to deal with vertical and horizontal animation separately?
* _Tentative_: Removing need for any CSS by using the [Web Animations API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API)?
  * Libraries like GSAP or Velocity are great for enabling pure JavaScript animation today (see: [`liquid-fire-velocity`](https://github.com/ember-animation/liquid-fire-velocity)), but they're a bit too heavy for this addon, which only seeks to provide a handful of base effects and mainly concerns itself with encapsulating functionality.


## Developing Locally

### Installation

* `git clone` this repository
* `npm install`
* `bower install`

### Running

* `ember server`
* Visit your app at http://localhost:4200.

### Running Tests

* `npm test` (Runs `ember try:testall` to test your addon against multiple Ember versions)
* `ember test`
* `ember test --server`

### Building

* `ember build`

For more information on using ember-cli, visit [http://www.ember-cli.com/](http://www.ember-cli.com/).



[demo]: http://www.sipple.io/ember-stagger-swagger-demo/


CHANGELOG.md
1.0.12 / 02-06-2016 (dd-mm-yyyy)
================================
* Fixed boolean logic when checking for an initial `outEffect` argument in `_resolveInitialEffectAttrs`. (`!!` was
  being used instead of `!`, causing an unnecessary warning message to be logged to the console.)

1.0.11 / 02-06-2016
================================
* Fixed an error in `_setAnimationValuesForItems` where `inDelay` and `outDelay` weren't being accounted for (thanks [@techsoldaten](https://github.com/techsoldaten))

1.0.10 / 15-05-2016
================================
* `visibility` is no longer set to `hidden` on items that have yet to be animated in (The only rule now applied is `opacity: 0`). This prevents an issue in Safari where even when toggling `visibility` back to `visible`, items would retain their status of `hidden`.

1.0.9 / 20-04-2016
================================
* Upgrade to `ember@2.5` and `ember-cli@2.6.0-beta.1`

1.0.8 / 20-04-2016
================================
* Fix typo in README describing when the `onAnimationStart` action was called.

1.0.7 / 09-04-2016
================================
* perform short-circuited default setting before warning for invalid arguments.
* use `Ember.Logger.warn` instead of `Ember.warn`

1.0.6 / 06-04-2016
================================
* hotfix: set `isDevelopingAddon` to false 😳

1.0.5 / 04-04-2016
================================
* Fix build by using Chrome in TravisCI.

1.0.4 / 03-04-2016
================================
* Remove dependency adding in blueprint (see: https://github.com/BrianSipple/ember-stagger-swagger/issues/1)

1.0.3 / 03-04-2016
================================
* Bring keyframes into main css file
  * Fixes broken import
  * Removes need for postCSS build step.

1.0.2 / 03-04-2016
================================
* Implement addPackagesToProject inside of the main blueprint `afterInstall` hook so that CSS is properly processed.

1.0.1 / 03-04-2016
================================
* Minor README fixes.

1.0.0 / 03-04-2016
================================
* Initial release version.


addon/mixins/stagger-list.js
/**
 * STAGGERED LIST
 *
 * A component mixin that renders list items by performing a staggered animation of their
 * entrance by listening to a "showItems" binding.
 *
 * NOTE: Right now, a "list item" is considered anything that's supplied as a direct child
 * element of the component's template block.
 */
import Ember from 'ember';
import setElementStyleProperty from 'ember-stagger-swagger/utils/set-element-style-property';
import getAnimationPrefix from 'ember-stagger-swagger/utils/get-animation-prefix';

const {
  Mixin,
  computed,
  run,
  warn,
  assert,
} = Ember;

const {
  notEmpty,
} = computed;


const defaults = {

  /**
   * 2 frames per item (1 frame @ 60fps ~= 16ms) creates a noticeably staggered
   * but still-perceptively fluid motion.
   * (see: https://en.wikipedia.org/wiki/Traditional_animation#.22Shooting_on_twos.22)
   */
  STAGGER_INTERVAL: 32,

  ANIMATION_DELAY: 0,

  TOTAL_DURATION_MS: 500,

  IN_TIMING_FUNCTION: 'cubic-bezier(0.215, 0.610, 0.355, 1.000)',  // ease-out-cubic
  OUT_TIMING_FUNCTION: 'cubic-bezier(0.55, 0.055, 0.675, 0.19)',  // ease-in-cubic
};


const CLASS_NAMES = {
  untoggled: 'hasnt-entered',
  listItemCompletedInitialToggle: 'completed-initial-enter',
  itemsHidden: 'items-hidden',
  itemsShowing: 'items-showing',
  itemsCollapsing: 'items-collapsing',
};

const ANIMATION_DIRECTIONS = {
  LEFT: 'left',
  DOWN: 'down',
  RIGHT: 'right',
  UP: 'up',
};

const ANIMATION_NAMES = {
  SLIDE_AND_FADE: 'slideAndFade',
  SLIDE: 'slide',
  FADE: 'fade',
  SCALE: 'scale',
};

const KEYFRAMES_MAP = {
  [ANIMATION_DIRECTIONS.LEFT]: {
    in: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeInFromRight',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideInFromRight',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeIn',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleIn',
    },
    out: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeOutLeft',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideOutLeft',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeOut',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleOut',
    },
  },
  [ANIMATION_DIRECTIONS.DOWN]: {
    in: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeInFromTop',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideInFromTop',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeIn',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleIn',
    },
    out: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeOutDown',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideOutDown',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeOut',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleOut',
    },
  },
  [ANIMATION_DIRECTIONS.RIGHT]: {
    in: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeInFromLeft',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideInFromLeft',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeIn',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleIn',
    },
    out: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeOutRight',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideOutRight',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeOut',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleOut',
    },
  },
  [ANIMATION_DIRECTIONS.UP]: {
    in: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeInFromBottom',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideInFromBottom',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeIn',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleIn',
    },
    out: {
      [ANIMATION_NAMES.SLIDE_AND_FADE]: '__EmberStaggerSwagger__SlideAndFadeOutUp',
      [ANIMATION_NAMES.SLIDE]: '__EmberStaggerSwagger__SlideOutUp',
      [ANIMATION_NAMES.FADE]: '__EmberStaggerSwagger__FadeOut',
      [ANIMATION_NAMES.SCALE]: '__EmberStaggerSwagger__ScaleOut',
    },
  },
};


export default Mixin.create({

  classNames: ['_ember-stagger-swagger_stagger-list'],
  classNameBindings: [`hasListToggled::${CLASS_NAMES.untoggled}`],

  /**
   * -------------------------------------------------------------------- *
   * ----------------------- API ------------------------ *
   * -------------------------------------------------------------------- *
   */

  /**
   * Flag for manually triggering either the show or hide animation
   *
   * By default, the list items will animate in on render, but
   * but this allows the user to have full toggle control if they want it.
   */
  showItems: true,

  /* trigger the entrance animation when this element is inserted into the DOM */
  enterOnRender: true,  // TODO: Support


  /* MILLESECONDS */
  staggerInterval: null,

  /* MILLESECONDS */
  animationDelay: null,

  // TODO: Support
  // inDelay: null,
  // outDelay: null,

  inDirection: null,
  outDirection: null,

  inEffect: null,
  outEffect: null,
  customInEffect: null,
  customOutEffect: null,


  /* Timing Function */
  inTimingFunc: null,
  outTimingFunc: null,
  timingFunc: null, // single timing function for both in and out

  /* Duration (milleseconds) */
  inDuration: 0,
  outDuration: 0,
  duration: 0,  // single duration for both in and out
  /**
   * -------------------------------------------------------------------- *
   * ----------------------- /API ------------------------ *
   * -------------------------------------------------------------------- *
   */

  isAnimating: false,
  hasListToggled: false,


  /**
   * Callback (to be initialized) for our animationstart event listener
   */
  _onStaggerStart: null,

  /**
   * Callback (to be initialized) for our animationend event listener
   */
  _onStaggerComplete: null,


  /**
   * Callback (to be initialized) for caching the DOM nodes of our child elements
   * when they get added to the DOM
   */
  _onChildElementsInserted: null,

  _childInsertionListener: null,

  /**
   * Array of cached "list item" elements to cache upon insertion
   */
  _listItemElems: null,

  /**
   * We'll be caching the animation prefix (if any) that's needed to
   * modify DOM element style properties.
   */
  _animationPrefix: null,


  hasItemsToAnimate: notEmpty('_listItemElems'),

  isReadyToAnimate: computed('hasItemsToAnimate', 'isAnimating', function isReadyToAnimate () {
    return this.get('hasItemsToAnimate') && !this.get('isAnimating');
  }),

  _inAnimationName: computed('inEffect', 'customInEffect', 'inDirection', function computeInAnimationName () {
    return this.get('customInEffect') || KEYFRAMES_MAP[this.get('inDirection')].in[this.get('inEffect')];
  }),

  _outAnimationName: computed('outEffect', 'customOutEffect', 'outDirection', function computeOutAnimationName () {
    return this.get('customOutEffect') || KEYFRAMES_MAP[this.get('outDirection')].out[this.get('outEffect')];
  }),

  currentAnimationName: computed(
    '_inAnimationName',
    '_outAnimationName',
    'showItems',
    'needsToAnimateAfterInit',
    function currentAnimationName() {
      return ( this.get('showItems') || this.get('needsToAnimateAfterInit') ) ?
        this.get('_inAnimationName')
        :
        this.get('_outAnimationName');
    }
  ),

  currentAnimationDuration: computed('duration', 'inDuration', 'outDuration', 'showItems', function computeDuration () {
    // give priority to a specified in/out duration
    if (this.get('duration')) {
      return this.get('duration');
    }

    // otherwise, set according to the state of showItems
    return this.get('showItems') ?
      this.get('inDuration') || defaults.TOTAL_DURATION_MS
      :
      this.get('outDuration') || defaults.TOTAL_DURATION_MS;
  }),

  currentAnimationTimingFunction: computed('inTimingFunc', 'outTimingFunc', 'timingFunc', 'showItems', function computeCurrentTimingFunction () {
    // give priority to a specified in/out duration
    if (this.get('timingFunc')) {
      return this.get('timingFunc');
    }

    // otherwise, set according to the state of showItems
    return this.get('showItems') ? this.get('inTimingFunc') : this.get('outTimingFunc');
  }),

  needsToAnimateAfterInit: computed('enterOnRender', 'hasListToggled', 'isReadyToAnimate', function needsInitialRender () {
    return (
      !this.get('hasListToggled') &&
      this.get('enterOnRender') &&
      this.get('isReadyToAnimate')
    );
  }),

  init () {
    this._super(...arguments);

    this._resolveInitialTimingAttrs();
    this._resolveInitialStaggerDirections();
    this._resolveInitialTimingFunctions();
  },


  didInsertElement () {
    this._super(...arguments);

    this._initAnimationCallbacks();

    if (this.element.children && this.element.children.length) {
      this._syncWithDOM();

    } else {
      this._initChildInsertionCallback();
      this._initChildInsertionListener();
    }
  },


  /**
   * Trigger the staggering animation when something on the outside updates `showItems`
   */
  didUpdateAttrs (attrData) {
    this._super(...arguments);
    const oldShowItems = (
      typeof attrData.oldAttrs.showItems.value !== 'undefined' &&
      attrData.oldAttrs.showItems.value
    );

    const newShowItems = (
      typeof attrData.newAttrs.showItems.value !== 'undefined' &&
      attrData.newAttrs.showItems.value
    );

    if (oldShowItems !== newShowItems && this.get('isReadyToAnimate')) {
      run.scheduleOnce('afterRender', this, '_triggerAnimation');
    }
  },


  willDestroyElement () {
    this._super(...arguments);

    const animationPrefix = this.get('animationPrefix');
    const startEvent = animationPrefix ? `${animationPrefix}AnimationStart` : 'animationstart';
    const endEvent = animationPrefix ? `${animationPrefix}AnimationEnd` : 'animationend';

    this.element.removeEventListener(`${startEvent}`, this._onStaggerStart, false);
    this.element.removeEventListener(`${endEvent}`, this._onStaggerComplete, false);

    this._listItemElems = null;
    this._childInsertionListener.disconnect();
  },



  _cacheListItems () {
    this.set('_listItemElems', Array.from(this.element.children));
  },

  _cacheAnimationPrefix () {
    this.set('_animationPrefix', getAnimationPrefix(this.element));
  },

  _initAnimationCallbacks () {

    this._onStaggerStart = function onStaggerStart (event) {
      this.set('isAnimating', true);
      this.send('broadcastAnimationStart', event);

    }.bind(this);

    /**
     * AnimationEvent listener for the `animationend` event fired by each
     * child item.
     *
     * If the event corresponds to the last item, we'll trigger a run loop call
     * that will fire immediately after the animation and
     * update `hasListToggled` (if necessary), `isAnimating`, and
     * tell our `broadcastAnimationComplete` action to fire
     *
     */
    this._onStaggerComplete = function onStaggerComplete (event) {

      // only update the DOM after we've finished animating all items
      const hasListToggled = this.get('hasListToggled');
      const lastListItemElem = this._listItemElems[this._listItemElems.length - 1];
      const targetElem = event.target;

      if (!hasListToggled) {
        targetElem.classList.add(CLASS_NAMES.listItemCompletedInitialToggle);
      }

      if (Object.is(targetElem, lastListItemElem)) {
        if (!hasListToggled) {
          this.set('hasListToggled', true);
        }
        run.once(() => {
          this.set('isAnimating', false);
          this.send('broadcastAnimationComplete', event);
        });
      }

    }.bind(this);
  },

  _syncWithDOM () {
    this._cacheListItems();
    this._cacheAnimationPrefix();
    this._prepareItemsInDOM();
    if (this.get('needsToAnimateAfterInit')) {
      run.scheduleOnce('afterRender', this, '_triggerAnimation');
      //this.set('_needsToAnimateAfterInit', true);
    }
  },

  _initChildInsertionCallback () {
    this._onChildElementsInserted = function onChildElementsInserted () {
      run.scheduleOnce('afterRender', this, () => {
        this._syncWithDOM();
      });
    }
  },

  /**
   * For components whose child elements are asynchronously rendered,
   * we can listen for the completion of such rendering and cache our
   * list items then.
   */
  _initChildInsertionListener () {
    this._childInsertionListener = new MutationObserver(function childInsertionListener (mutations) {
      this._onChildElementsInserted();
    }.bind(this));

    this._childInsertionListener.observe(this.element, { childList: true });
  },

  _prepareItemsInDOM () {
    this._setAnimationValuesForItems(false);

    const animationPrefix = this.get('_animationPrefix');
    const startEvent = animationPrefix ? `${animationPrefix}AnimationStart` : 'animationstart';
    const endEvent = animationPrefix ? `${animationPrefix}AnimationEnd` : 'animationend';

    this.element.addEventListener(`${startEvent}`, this._onStaggerStart, false);
    this.element.addEventListener(`${endEvent}`, this._onStaggerComplete, false);
  },


  _resolveInitialTimingAttrs () {
    if (!this.staggerInterval) {
      this.staggerInterval = defaults.STAGGER_INTERVAL;

    } else {

      /* eslint-disable max-len */
      warn(
        `The stagger interval that you attempted to specify was invalid. Please use a numeric value greater than 0. Defaulting to ${defaults.STAGGER_INTERVAL}`,
        !Number.isNaN(Number(this.staggerInterval)) && this.staggerInterval > 0
      );
      /* eslint-enable max-len */
    }

    if (!this.animationDelay) {
      this.animationDelay = defaults.ANIMATION_DELAY;

    } else {

      /* eslint-disable max-len */
      warn(
        `The initial delay that you attempted to specify was invalid. Please use a numeric value. Defaulting to ${defaults.ANIMATION_DELAY}`,
        !Number.isNaN(Number(this.staggerInterval))
      );
      /* eslint-enable max-len */
    }
  },

  _resolveInitialStaggerDirections () {
    // enforce inDirection
    assert(
      'stagger-list must have a valid `inDirection`',
      !!this.inDirection && !!KEYFRAMES_MAP[this.inDirection]
    );

    // if not set, set the default outDirection to the continuation of the inDirection
    if (!this.outDirection) {
      this.outDirection = this.inDirection;

    } else if (!KEYFRAMES_MAP[this.outDirection]) {
      warn(`invalid \`outDirection\`. Defaulting to the continuation of \`inDirection\`
        https://https://github.com/BrianSipple/ember-stagger-swagger#usage`
      );
      this.outDirection = this.inDirection;
    }
  },

  _resolveInitialTimingFunctions () {
    this.inTimingFunc = this.inTimingFunc || defaults.IN_TIMING_FUNCTION;
    this.outTimingFunc = this.outTimingFunc || defaults.OUT_TIMING_FUNCTION;
  },


  _setAnimationValuesForItems (applyName = false) {

    // setting an animation name is what we use to trigger the animation, so only set if asked for
    const currentAnimationName = applyName ? this.get('currentAnimationName'): '';
    const currentStaggerInterval = this.get('staggerInterval');
    const currentAnimationDuration = `${this.get('currentAnimationDuration')}ms`;
    const currentAnimationTimingFunction = this.get('currentAnimationTimingFunction');

    const animationPrefix = this.get('_animationPrefix');
    const propertyPrefix = animationPrefix ? `${animationPrefix}A` : 'a';

    this._listItemElems.forEach((listItemElem, idx) => {
      listItemElem.style[`${propertyPrefix}nimationDelay`] = `${currentStaggerInterval * (idx + 1)}ms`;
      listItemElem.style[`${propertyPrefix}nimationTimingFunction`] = currentAnimationTimingFunction;
      listItemElem.style[`${propertyPrefix}nimationDuration`] = currentAnimationDuration;
      listItemElem.style[`${propertyPrefix}nimationName`] = currentAnimationName;
      listItemElem.style[`${propertyPrefix}nimationFillMode`] = 'both';
    });
  },

  _triggerAnimation() {
    this._setAnimationValuesForItems(true);
  },

  actions: {

    broadcastAnimationStart(animationEvent) {
      if (typeof this.onAnimationStart === 'function') {
        this.onAnimationStart(animationEvent);
      }
    },

    broadcastAnimationComplete(animationEvent) {
      if (typeof this.onAnimationComplete === 'function') {
        this.onAnimationComplete(animationEvent);
      }
    },
  }
});
